docs/beyond_basics/101-168-dvc.rst

```diff
@@ -0,0 +1,926 @@
+   All tools are available via `pip <https://pypi.org/project/pip/>`_ or `conda <https://docs.conda.io/en/latest/>`_.
+   We recommend to install them in a `virtual environment <https://realpython.com/python-data-version-control/#set-up-your-working-environment>`_ -- the DVC tutorial has `step-by-step instructions <https://realpython.com/python-data-version-control/#set-up-your-working-environment>`_.
+
+Version controlling data
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+In the first part of the tutorial, the directory tree will be populated with data that should be version controlled.
+
+Although the implementation of version control for (large) data is very different between DataLad and DVC, the underlying concept is very similar:
+(Large) data is stored outside of :term:`Git` -- :term:`Git` only tracks information on where this data can be found.
+
+In DataLad datasets, (large) data is handled by :term:`git-annex`.
+Data content is `hashed <https://en.wikipedia.org/wiki/Hash_function>`_ and only the hash (represented as the original file name) is stored in Git [#f2]_.
+Actual data is stored in the :term:`annex` of the dataset, and annexed data can be transferred from and to a `large number of storage solutions <https://git-annex.branchable.com/special_remotes/>`_ using either DataLad or git-annex commands.
+Information on where data is available from is :ref:`stored in an internal representation of git-annex <gitannexbranch>`.
+
+In DVC repositories, (large) data is also supposed to be stored in external remotes such as Google Drive.
+But instead of an internal representation of where files are available from, DVC repositories have one ``.dvc`` text file for each data file or directory given to DVC.
+The ``.dvc`` files contain information on the path to the data in the repository, where the associated data file is available from, and a hash, and users are supposed to store them in :term:`Git`.
+
+
+.. findoutmore:: Download data for coding along
+
+   The data used in the tutorial is the "Imagenette" dataset and can be `downloaded from an Amazon S3 bucket <https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-160.tgz>`_.
+   Download the data and extract it into the ``data/raw/`` directory in both repositories.
+
+   .. runrecord:: _examples/DL-101-168-108
+      :workdir: DVCvsDL
+      :language: console
+
+      # download the data
+      $ curl -s https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-160.tgz \
+                -O imagenette2-160.tgz
+      # extract it
+      $ tar -xzf imagenette2-160.tgz
+      # move it into the directories in both repositories
+      $ cp -r imagenette2-160/train DVC-DataLad/data/raw/
+      $ cp -r imagenette2-160/val DVC-DataLad/data/raw/
+      $ cp -r imagenette2-160/train DVC/data/raw/
+      $ cp -r imagenette2-160/val DVC/data/raw/
+      # remove the archive and extracted folder
+      $ rm -rf imagenette2-160
+      $ rm imagenette2-160.tgz
```

Comment:
Would it make sense to mention that we could pull the data straight with datalad?

Reply:
I've thought about doing this, but then decided to stay procedurally close to the DVC workflow - but, yes, we could definitely mention it

Reply:
Since you want a "comparison" I think it would be beneficial to show how to do it with DataLad functionality.  This component is really a part of the "DVC workflow".  And you do give examples of "how to do with DVC" so I would have moved this into "DVC" section below AND below in "DataLad workflow" section I would have started with a new dataset (instead of if I got it right just adding datalad on top of what is downloaded here for DVC way) and used `datalad download-url --archive ...` (may be followed with `git mv`?) to demonstrate how easy it is to  get data from external tarballs while keeping the access to them.  Without that I also (may be missed) not sure how you in DataLad workflow add ability to redownload all these input files for reproducibility.

That might also appeal to DVC developers to may be think (again) about integrating/supporting datalad.

Reply:
Thanks for your thoughts - yes, true, I think you convinced me. Will change this.

Reply:
Mhhh, what's the most elegant way of downloading the data? 
``datalad download-url --archive 'https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-160' -O 'data/raw/`` results in a downloaded, annexed tar ball:

```
├── data
│   ├── prepared
│   └── raw
│       └── MD5E-s98948031--0edfc972b5c9817ac36517c0057f3869.tar -> ../../.git/annex/objects/32/Px/MD5E-s116828160--1e92a6d123bfaf23515ecbb3dc221cad.tar/MD5E-s116828160--1e92a6d123bfaf23515ecbb3dc221cad.tar
```
``tar -xf``'ing the tarball gets me the extracted data, but untracked. I could move it and save it, but I end up with the tarball and the data in the dataset.I *thought* (but maybe I misunderstood the command so far) that `add-archive-content` would extract my files for me and keep the tarball in a dedicated branch?

Reply:
@yarikoptic do you see a better way than

```
$ datalad download-url --archive 'https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-160.tgz' -O 'data/raw/.tarballarchive'
$ cd data/raw && datalad add-archive-content --strip-leading-dirs .tarballarchive
```
?

And (sorry for my ignorance here, but I haven't understood add-archive-content fully yet) could you check whether my assumptions here are correct? As I understood it:
-  the original tarball (``.tarballarchive``) and the extracted data from it is currently in the dataset
- the s3 URL origin of the tarball is also registered as a special remote
- I could drop the tarball
- If I would need to reobtain a single extracted file, datalad would re-download the tarball and get it from there

Reply:
I think there is a bug in datalad (re)implementation of functionality we used from patool: https://github.com/datalad/datalad/issues/4875

workaround -- use patool:
```
datalad -c datalad.runtime.use-patool=1 download-url --archive 'https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-160.tgz' -O 'data/raw/'
```

<details>
<summary>output (takes AWHILE to add all those files without any progress indication: https://github.com/datalad/datalad/issues/4876)</summary> 

```shell
$> datalad -c datalad.runtime.use-patool=1 download-url --archive 'https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-160.tgz' -O 'data/raw/'
[INFO   ] Downloading 'https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-160.tgz' into '/tmp/testds/data/raw/' 
download_url(ok): /tmp/testds/data/raw/imagenette2-160.tgz (file)                                                
add(ok): /tmp/testds/data/raw/imagenette2-160.tgz (file)
save(ok): /tmp/testds (dataset)
[INFO   ] Adding content of the archive /tmp/testds/data/raw/imagenette2-160.tgz into annex AnnexRepo(/tmp/testds) 
[INFO   ] Finished adding /tmp/testds/data/raw/imagenette2-160.tgz: Files processed: 13394, renamed: 13394, +annex: 13394 
action summary:
  add (ok: 1)
  download_url (ok: 1)
  save (ok: 1)
datalad -c datalad.runtime.use-patool=1 download-url --archive  -O 'data/raw/  190.24s user 32.21s system 106% cpu 3:28.88 total
```
</details>

if you check `git log` you will see that there are two commits from this operation -- one adds archive to git/git-annex, 2nd one extracts and removes that tarball.  You could drop tarball if so desired, but you would need to either know checksum, or checkout that commit and drop, or use `git annex unused && git annex drop --unused` tandem:

<details>
<summary>drop unused (.tar was there from previous attempts without patool)</summary> 

```shell
$> du -scm .git/annex
449	.git/annex
449	total
(dev3) 1 10485.....................................:Sat 05 Sep 2020 10:45:59 AM EDT:.
(git)smaug:/tmp/testds[master]git
$> git annex unused && git annex drop --unused
unused . (checking for unused data...) (checking master...) 
  Some annexed data is no longer used by any files:
    NUMBER  KEY
    1       MD5E-s116828160--1e92a6d123bfaf23515ecbb3dc221cad.tar
    2       MD5E-s98948031--0edfc972b5c9817ac36517c0057f3869.tgz
  (To see where data was previously used, try: git log --stat --no-textconv -S'KEY')
  
  To remove unwanted data: git-annex dropunused NUMBER
  
ok
drop MD5E-s98948031--0edfc972b5c9817ac36517c0057f3869.tgz (checking https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-160.tgz...) ok
drop MD5E-s116828160--1e92a6d123bfaf23515ecbb3dc221cad.tar ok
(recording state in git...)
(dev3) 1 10486.....................................:Sat 05 Sep 2020 10:46:13 AM EDT:.
(git)smaug:/tmp/testds[master]git
$> du -scm .git/annex                         
243	.git/annex
243	total

```
</details>

Reply:
forgot to add listing to show that it worked:
```
$> ls data/raw/imagenette2-160/*    
data/raw/imagenette2-160/train:
n01440764/  n02979186/  n03028079/  n03417042/  n03445777/
n02102040/  n03000684/  n03394916/  n03425413/  n03888257/

data/raw/imagenette2-160/val:
n01440764/  n02979186/  n03028079/  n03417042/  n03445777/
n02102040/  n03000684/  n03394916/  n03425413/  n03888257/
```

Reply:
Thank you!

Reply:
Mhhh, getting another interesting error. When I try to push this dataset to a sibling created with "datalad create sibling" I can't push the git-annex branch:

```
datalad push --to my-sibling
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00000293.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00002138.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00003014.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00006697.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00007197.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00009346.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00009379.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00009396.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00010306.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00011233.JPEG (file) [to my-sibling...]         
Update availability for 'my-sibling':  75%|████████████████████████        | 3.00/4.00 [00:00<00:00, 29.5k Steps/s]CommandError: 'git push --progress --porcelain my-sibling refs/heads/git-annex:refs/heads/git-annex git-annex:git-annex' failed with exitcode 1 under /tmp/DVC-DataLad
error: dst ref refs/heads/git-annex receives from more than one src.
error: failed to push some refs to '/tmp/datalad-sibling'
```

<details> <summary>Here is what I have ran:</summary>

```
$ datalad create -c text2git -c yoda DVC-DataLad
$ cd DVC-DataLad && mkdir -p data/{raw,prepared} model metrics
$ datalad download-url -m "download scripts for ML analysis" \
     https://github.com/datalad-handbook/data-version-control/blob/master/src/{train,prepare,evaluate}.py \
     -O 'code/'
$ datalad -c datalad.runtime.use-patool=1 download-url --archive 'https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-160.tgz' -O 'data/raw/'
$ mv data/raw/imagenette2-160/* data/raw/ && rmdir data/raw/imagenette2-160
$ datalad save -m "Move data into preferred locations"
$ datalad create-sibling --name my-sibling ../datalad-sibling
$ datalad push --to my-sibling
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00000293.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00002138.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00003014.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00006697.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00007197.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00009346.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00009379.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00009396.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00010306.JPEG (file) [to my-sibling...]         
copy(ok): /tmp/DVC-DataLad/data/raw/train/n01440764/ILSVRC2012_val_00011233.JPEG (file) [to my-sibling...]         
Update availability for 'my-sibling':  75%|████████████████████████        | 3.00/4.00 [00:00<00:00, 29.5k Steps/s]CommandError: 'git push --progress --porcelain my-sibling refs/heads/git-annex:refs/heads/git-annex git-annex:git-annex' failed with exitcode 1 under /tmp/DVC-DataLad
error: dst ref refs/heads/git-annex receives from more than one src.
error: failed to push some refs to '/tmp/datalad-sibling'

```

</details>

This also happens when I clone the dataset and add the clone as a sibling.

Reply:
oh look, the error message seems to suggest an interesting issue: https://git-annex.branchable.com/forum/refs__47__heads__47__synced__47__git-annex_receives_from_more_than_one_src/

Reply:
I filed https://github.com/datalad/datalad/issues/4888. I think we can express the handbook contribution to DataLad in "bugs per section" :wink: